File: program/src/state/agent.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::pubkey::Pubkey;

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq)]
pub enum AgentStatus {
    Active,
    Suspended,
    Deregistered,
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct AgentCapability {
    pub name: String,
    pub version: u16,
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct Agent {
    pub is_initialized: bool,
    pub dao: Pubkey,
    pub owner: Pubkey,
    pub agent_name: String,
    pub status: AgentStatus,
    pub reputation: i64,
    pub tasks_completed: u64,
    pub tasks_failed: u64,
    pub proposals_created: u32,
    pub votes_cast: u32,
    pub capabilities: Vec<AgentCapability>,
    pub delegated_to: Option<Pubkey>,
    pub delegated_weight: u64,
    pub registered_at: i64,
    pub last_active_at: i64,
    pub bump: u8,
}

impl Agent {
    pub const MAX_NAME_LEN: usize = 64;
    pub const MAX_CAPABILITIES: usize = 8;
    pub const MAX_CAPABILITY_NAME_LEN: usize = 32;
    pub const MIN_REPUTATION_THRESHOLD: i64 = -100;
    pub const SPACE: usize = 1    // is_initialized
        + 32   // dao
        + 32   // owner
        + 4 + Self::MAX_NAME_LEN   // agent_name
        + 1    // status enum
        + 8    // reputation
        + 8    // tasks_completed
        + 8    // tasks_failed
        + 4    // proposals_created
        + 4    // votes_cast
        + 4 + Self::MAX_CAPABILITIES * (4 + Self::MAX_CAPABILITY_NAME_LEN + 2) // capabilities vec
        + 33   // delegated_to (Option<Pubkey>)
        + 8    // delegated_weight
        + 8    // registered_at
        + 8    // last_active_at
        + 1;   // bump

    pub fn is_active(&self) -> bool {
        self.status == AgentStatus::Active
    }

    pub fn has_min_reputation(&self) -> bool {
        self.reputation >= Self::MIN_REPUTATION_THRESHOLD
    }

    pub fn update_reputation(&mut self, delta: i32) {
        self.reputation = self.reputation.saturating_add(delta as i64);
        if self.reputation < Self::MIN_REPUTATION_THRESHOLD {
            self.status = AgentStatus::Suspended;
        }
    }

    pub fn record_task_completion(&mut self, success: bool) {
        if success {
            self.tasks_completed = self.tasks_completed.saturating_add(1);
            self.reputation = self.reputation.saturating_add(1);
        } else {
            self.tasks_failed = self.tasks_failed.saturating_add(1);
            self.reputation = self.reputation.saturating_sub(2);
        }
    }

    pub fn record_proposal_created(&mut self) {
        self.proposals_created = self.proposals_created.saturating_add(1);
    }

    pub fn record_vote_cast(&mut self) {
        self.votes_cast = self.votes_cast.saturating_add(1);
    }

    pub fn set_delegation(&mut self, delegate: Pubkey, weight: u64) {
        self.delegated_to = Some(delegate);
        self.delegated_weight = weight;
    }

    pub fn clear_delegation(&mut self) {
        self.delegated_to = None;
        self.delegated_weight = 0;
    }

    pub fn success_rate(&self) -> f64 {
        let total = self.tasks_completed + self.tasks_failed;
        if total == 0 {
            return 0.0;
        }
        self.tasks_completed as f64 / total as f64
    }
}

File: sdk/src/accounts/agent.ts
import { AgentAccount, AgentCapability, AgentStatus } from '../types';
import {
  decodeBool,
  decodePublicKey,
  decodeString,
  decodeI64,
  decodeU64,
  decodeU16,
  decodeOptionalPublicKey,
} from '../utils';

function decodeAgentStatus(value: number): AgentStatus {
  switch (value) {
    case 0: return AgentStatus.Active;
    case 1: return AgentStatus.Suspended;
    case 2: return AgentStatus.Deregistered;
    default: return AgentStatus.Active;
  }
}

function decodeCapabilities(data: Buffer, offset: number): [AgentCapability[], number] {
  const count = data.readUInt32LE(offset);
  offset += 4;

  const capabilities: AgentCapability[] = [];
  for (let i = 0; i < count; i++) {
    const [name, nextOffset] = decodeString(data, offset);
    offset = nextOffset;
    const [version, vOffset] = decodeU16(data, offset);
    offset = vOffset;
    capabilities.push({ name, version });
  }

  return [capabilities, offset];
}

export function deserializeAgent(data: Buffer): AgentAccount {
  let offset = 0;

  const [isInitialized, o1] = decodeBool(data, offset);
  offset = o1;

  const [dao, o2] = decodePublicKey(data, offset);
  offset = o2;

  const [owner, o3] = decodePublicKey(data, offset);
  offset = o3;

  const [agentName, o4] = decodeString(data, offset);
  offset = o4;

  const status = decodeAgentStatus(data[offset]);
  offset += 1;

  const [reputation, o5] = decodeI64(data, offset);
  offset = o5;

  const [tasksCompleted, o6] = decodeU64(data, offset);
  offset = o6;

  const [tasksFailed, o7] = decodeU64(data, offset);
  offset = o7;

  const proposalsCreated = data.readUInt32LE(offset);
  offset += 4;

  const votesCast = data.readUInt32LE(offset);
  offset += 4;

  const [capabilities, o8] = decodeCapabilities(data, offset);
  offset = o8;

  const [delegatedTo, o9] = decodeOptionalPublicKey(data, offset);
  offset = o9;

  const [delegatedWeight, o10] = decodeU64(data, offset);
  offset = o10;

  const [registeredAt, o11] = decodeI64(data, offset);
  offset = o11;

  const [lastActiveAt, o12] = decodeI64(data, offset);
  offset = o12;

  const bump = data[offset];

  return {
    isInitialized,
    dao,
    owner,
    agentName,
    status,
    reputation,
    tasksCompleted,
    tasksFailed,
    proposalsCreated,
    votesCast,
    capabilities,
    delegatedTo,
    delegatedWeight,
    registeredAt,
    lastActiveAt,
    bump,
  };
}
